src/xma/src/xmaapi/xmakernel.cpp

```diff
@@ -240,39 +195,57 @@ xma_kernel_session_create(XmaKernelProperties *props)
     priv1->kernel_execbos.reserve(num_execbo);
     priv1->num_execbo_allocated = num_execbo;
     if (xma_core::create_session_execbo(priv1, num_execbo, XMA_KERNEL_MOD) != XMA_SUCCESS) {
-        //Release singleton lock
-        g_xma_singleton->locked = false;
         free(session->base.plugin_data);
         free(session);
         delete priv1;
         return nullptr;
     }
 
-    rc = session->kernel_plugin->init(session);
-    if (rc) {
-        xma_logmsg(XMA_ERROR_LOG, XMA_KERNEL_MOD,
-                   "Initalization of kernel plugin failed. Return code %d\n",
-                   rc);
-        //Release singleton lock
-        g_xma_singleton->locked = false;
-        free(session->base.plugin_data);
-        free(session);
-        delete priv1;
-        return nullptr;
+    //Obtain lock only for a) singleton changes & b) kernel_info changes
+    std::unique_lock<std::mutex> guard1(g_xma_singleton->m_mutex);
+    //Singleton lock acquired
+
+    if (!kernel_info->soft_kernel && !kernel_info->in_use && !kernel_info->context_opened) {
+        if (xclOpenContext(dev_handle, dev_tmp1.uuid, kernel_info->cu_index_ert, true) != 0) {
+            xma_logmsg(XMA_ERROR_LOG, XMA_KERNEL_MOD, "Failed to open context to CU %s for this session\n", kernel_info->name);
+            free(session->base.plugin_data);
+            free(session);
+            delete priv1;
+            return nullptr;
+        }
     }
+    session->base.session_id = g_xma_singleton->num_of_sessions + 1;
+    xma_logmsg(XMA_INFO_LOG, XMA_KERNEL_MOD,
+                "XMA session channel_id: %d; session_id: %d\n", session->base.channel_id, session->base.session_id);
 
     if (kernel_info->in_use) {
         kernel_info->is_shared = true;
+        xma_logmsg(XMA_DEBUG_LOG, XMA_KERNEL_MOD,
+                   "XMA session sharing CU: %s\n", hwcfg->devices[hwcfg_dev_index].kernels[cu_index].name);
     } else {
         kernel_info->in_use = true;
+        xma_logmsg(XMA_DEBUG_LOG, XMA_KERNEL_MOD,
+                   "XMA session with CU: %s\n", hwcfg->devices[hwcfg_dev_index].kernels[cu_index].name);
     }
     kernel_info->num_sessions++;
     g_xma_singleton->num_kernels++;
     g_xma_singleton->num_of_sessions = session->base.session_id;
-    g_xma_singleton->all_sessions.emplace(g_xma_singleton->num_of_sessions, session->base);
+    g_xma_singleton->all_sessions_vec.push_back(session->base);
 
     //Release singleton lock
-    g_xma_singleton->locked = false;
+    guard1.unlock();
+
+    //init can execute cu cmds as well so must be fater adding to singleton above
+    rc = session->kernel_plugin->init(session);
+    if (rc) {
+        xma_logmsg(XMA_ERROR_LOG, XMA_KERNEL_MOD,
+                   "Initalization of kernel plugin failed. Return code %d\n",
+                   rc);
+        free(session->base.plugin_data);
+        //free(session); Added to singleton above; Keep it as checked for cu cmds
```

Comment:
Why is the session not being freeded or the "plugin_data" value not being set to a nullptr.